core/src/Instance/B2CAuthority.cs

```diff
@@ -37,37 +38,85 @@ internal class B2CAuthority : AadAuthority
 
-        internal B2CAuthority(string authority, bool validateAuthority) 
+        internal B2CAuthority(string authority, bool validateAuthority)
             : base(authority, validateAuthority)
         {
+            AuthorityType = AuthorityType.B2C;
+            ValidateAuthority = validateAuthority;
+
             Uri authorityUri = new Uri(authority);
-            string[] pathSegments = authorityUri.AbsolutePath.Substring(1).Split(new [] { '/'}, StringSplitOptions.RemoveEmptyEntries);
+            string[] pathSegments = authorityUri.AbsolutePath.Substring(1).Split(new[] { '/' }, StringSplitOptions.RemoveEmptyEntries);
             if (pathSegments.Length < 3)
             {
                 throw new ArgumentException(CoreErrorMessages.B2cAuthorityUriInvalidPath);
             }
 
             CanonicalAuthority = string.Format(CultureInfo.InvariantCulture, B2CCanonicalAuthorityTemplate, authorityUri.Authority,
                 pathSegments[0], pathSegments[1], pathSegments[2]);
-
-            AuthorityType = AuthorityType.B2C;
         }
 
-        protected override async Task<string> GetOpenIdConfigurationEndpointAsync(
-            IHttpManager httpManager, 
-            ITelemetryManager telemetryManager, 
-            string userPrincipalName, 
+        internal override async Task UpdateCanonicalAuthorityAsync(
+            IHttpManager httpManager,
+            ITelemetryManager telemetryManager,
             RequestContext requestContext)
         {
-            if (ValidateAuthority && !IsInTrustedHostList(new Uri(CanonicalAuthority).Host))
+            if (ValidateAuthority && !InTrustedHostList(new Uri(CanonicalAuthority).Host))
             {
                 throw new ArgumentException(CoreErrorMessages.UnsupportedAuthorityValidation);
             }
 
+            var metadata = await AadInstanceDiscovery
+                             .Instance.GetMetadataEntryAsync(
+                                 httpManager,
+                                 telemetryManager,
+                                 new Uri(UpdateCanonicalAuthorityB2C()),
+                                 ValidateAuthority,
+                                 requestContext)
+                             .ConfigureAwait(false);
+
+            CanonicalAuthority = UpdateHost(CanonicalAuthority, metadata.PreferredNetwork);
+        }
+
+        private string UpdateCanonicalAuthorityB2C()
+        {
+            Uri b2cAuthority = new Uri(CanonicalAuthority);
+
+            string canonicalAuthorityUri = string.Format(CultureInfo.InvariantCulture, MicrosoftOnline + b2cAuthority.AbsolutePath);
+            return canonicalAuthorityUri;
+        }
+
+        private bool InTrustedHostList(string host)
+        {
+            bool isInList =
+                !string.IsNullOrEmpty(
+                    TrustedHostList.FirstOrDefault(a => string.Compare(host, a, StringComparison.OrdinalIgnoreCase) == 0));
```

Comment:
consider replacing this with the method which already exists on AadAuthority. Also you should probably combine the two comparisons into one "if" statement to make it conditional if both conditions should be evaluated (it will have a few CPU cycles).

Reply:
yes...good feedback. thanks.

---
In reply to: [231014061](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1383#discussion_r231014061) [](ancestors = 231014061)